Monika-After-Story/MonikaModDev
# Update Marisa

# TODO
- [x] thumbnail for witch hat
- [x] select dlg for witch hat
- [x] additional disable dlg for `DISB_ACS_BC_HAIR` if desired

Marisa is now equipped to the current sprite system (at long last)

Miku is now spritepacked, and o31's Miku greeting has been adjusted accordingly.

## Other affected released spritepacks:
* Sakuya - o31 added to costume type
* Shell Bikini - `bare-right-shoulder` added to exprops
* Shoulderless Sweater - `bare-right-shoulder` added to exprops
* Pink Shirt - `bare-right-shoulder` added to exprops

## Additional changes
* `mas_selspr.startup_prompt_check` - new function to correctly set prompts for selectors
* new acs type `hat` with selector for all hat-based ACS
* ribbon reactions now should not force outfit change to default if wearing Marisa. 
* `DISB_ACS_BC_HAIR` - `disable_type` constant for ACS that need to be disabled because of hair
* `mas_selspr.is_hairacs_compatible` - wrapper around `mas_sprites.is_hairacs_compatible` that can compare a hair object to an ACS selectable
* `mas_selspr.is_clotheshair_compatible` - wrapper around `mas_sprites.is_clotheshair_compatible` that can compare a clothes object to a hair selectable
* `mas_selspr.set_compat_acs` - sets `disable_type` for a list of ACS selectables based on given hair sprite object
* `mas_selspr.set_compat_hair` - sets `disable_type` for a list of hair selectables based on given clothes sprite object
* fixed `_history_list` crashes in selector sidebar
* `mas_selector_generic_sidebar_select_acs` - label that generalizes common ACS selector code. Mainly for ACS that do not have complicated filtering or dialogue.

## New exprops (and constants)
* `EXP_H_TS` - hair ex prop that denotes a hairstyle is a `tiedstrand` style
* `EXP_H_NT` - hair ex prop that denotes a hairstyle has no "tails" (ponytail/twintail/bun/etc...)
* `EXP_C_COST` - new constant for `costume` ex prop
* `EXP_C_COSP` - new constant for `cosplay` ex prop
* `EXP_C_LING` - new constant for `lingerie` ex prop
* `DEF_MUX_HAT` - default mux types for hats

## Testing:
* Verify that Miku is still unlocked if you got it on o31 2019
* Verify that Miku still looks correct
* Verify that Marisa looks good
* Verify that Marisa, the Marisa hair, and the Witch hat all unlock via update scripts if the user saw the costume before. (**NOTE: it would be wise to also test the main unlock flow, but we can do that in october**)
* Verify when switching out of Marisa:
    * the strandbow is always removed
    * the witch hat is only removed if outfit mode is enabled
* Verify Down Tied Hair Strand (Marisa hair) is disabled in the hair selector if Monika is wearing clothes that do not have `bare-right-shoulder`.
* Verify that if given a ribbon for the first time while wearing Marisa, the outfit is **not** changed to uniform. **Hairstyle should change to default if not an appropriate ribbon hairstyle**
* Verify hat selector is only shown when a hat is unlocked (and at happy+ affection)
* Verify Witch Hat is disabled in the hat selector if Monika is wearing a hairstyle with tails (like ponytail/twintail/etc...)
* verify no `_history_list` crashes when wildly hovering/clicking selectables in a selector screen.

### regression testing
* verify that all selector prompts are correctly set on startup ("change" used when currently wearing an ACS, "wear" used when not)
* verify that all selector topics are working correctly (all unlocked selectors are shown)
